File: mcop_package/mcop/integrations/haystack.py
"""MCOP ↔ Haystack integration shim (Python).

Implements the Haystack 2.x ``DocumentStore`` protocol shape without
taking a runtime dependency on ``haystack-ai``.

Mirrors `src/integrations/haystack.ts` 1:1.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from typing import Any, Dict, List, Literal, Optional, Sequence

from .triad_harness import (
    MCOPProvenance,
    MCOPTriad,
    MCOPTriadOptions,
    ensure_triad,
    recall_from_triad,
    record_into_triad,
)

__all__ = [
    "MCOPHaystackDocument",
    "MCOPHaystackDocumentStore",
    "create_mcop_haystack_document_store",
    "mcop_haystack_document_from_content",
]

MCOPHaystackDuplicatePolicy = Literal["overwrite", "skip", "fail"]


@dataclass(frozen=True)
class MCOPHaystackDocument:
    id: str
    content: str
    meta: Optional[Dict[str, Any]] = None
    score: Optional[float] = None
    embedding: Optional[Sequence[float]] = None
    provenance: Optional[MCOPProvenance] = None


@dataclass
class MCOPHaystackDocumentStore:
    """MCOP-backed Haystack 2.x ``DocumentStore`` implementation."""

    default_policy: MCOPHaystackDuplicatePolicy = "overwrite"
    triad_options: Optional[MCOPTriadOptions] = None
    _documents: Dict[str, MCOPHaystackDocument] = field(default_factory=dict)
    _docs_by_trace_id: Dict[str, str] = field(default_factory=dict)
    _triad: Optional[MCOPTriad] = None

    def __post_init__(self) -> None:
        self._triad = ensure_triad(self.triad_options)

    def count_documents(self) -> int:
        return len(self._documents)

    def write_documents(
        self,
        documents: Sequence[MCOPHaystackDocument],
        policy: Optional[MCOPHaystackDuplicatePolicy] = None,
    ) -> int:
        effective = policy or self.default_policy
        triad = self._require_triad()
        written = 0
        for doc in documents:
            existing = self._documents.get(doc.id)
            if existing is not None:
                if effective == "fail":
                    raise ValueError(f"Duplicate document id: {doc.id}")
                if effective == "skip":
                    continue
            recorded = record_into_triad(
                triad,
                doc.content,
                metadata={
                    **(doc.meta or {}),
                    "mcop_haystack_document_id": doc.id,
                },
                note=f"mcop-haystack:{doc.id}",
            )
            stored = MCOPHaystackDocument(
                id=doc.id,
                content=doc.content,
                meta={
                    **(doc.meta or {}),
                    "mcop_stigmergy_trace_id": recorded.trace.id,
                    "mcop_etch_hash": recorded.etch.hash,
                },
                score=doc.score,
                embedding=doc.embedding or recorded.trace.context,
                provenance=recorded.provenance,
            )
            self._documents[stored.id] = stored
            self._docs_by_trace_id[recorded.trace.id] = stored.id
            written += 1
        return written

    def filter_documents(
        self, filters: Optional[Dict[str, Any]] = None
    ) -> List[MCOPHaystackDocument]:
        all_docs = list(self._documents.values())
        if not filters:
            return all_docs
        return [
            doc
            for doc in all_docs
            if all((doc.meta or {}).get(k) == v for k, v in filters.items())
        ]

    def delete_documents(self, document_ids: Sequence[str]) -> None:
        for doc_id in document_ids:
            self._documents.pop(doc_id, None)
            for trace_id, mapped_id in list(self._docs_by_trace_id.items()):
                if mapped_id == doc_id:
                    del self._docs_by_trace_id[trace_id]

    def recall_by_resonance(self, query: str) -> Dict[str, Any]:
        triad = self._require_triad()
        _, resonance = recall_from_triad(triad, query)
        if resonance.trace is None:
            return {"score": resonance.score, "document": None}
        doc_id = self._docs_by_trace_id.get(resonance.trace.id)
        document = self._documents.get(doc_id) if doc_id is not None else None
        return {"score": resonance.score, "document": document}

    @property
    def triad_handle(self) -> MCOPTriad:
        return self._require_triad()

    def _require_triad(self) -> MCOPTriad:
        if self._triad is None:
            self._triad = ensure_triad(self.triad_options)
        return self._triad


def create_mcop_haystack_document_store(
    default_policy: MCOPHaystackDuplicatePolicy = "overwrite",
    triad_options: Optional[MCOPTriadOptions] = None,
) -> MCOPHaystackDocumentStore:
    return MCOPHaystackDocumentStore(
        default_policy=default_policy, triad_options=triad_options
    )


def mcop_haystack_document_from_content(
    content: str, meta: Optional[Dict[str, Any]] = None
) -> MCOPHaystackDocument:
    return MCOPHaystackDocument(id=str(uuid.uuid4()), content=content, meta=meta)

File: src/integrations/haystack.ts
/**
 * MCOP ↔ Haystack integration shim.
 *
 * Implements the Haystack 2.x `DocumentStore` protocol shape without
 * taking a runtime dependency on `haystack-ai`. The Python-canonical
 * sibling lives in `mcop_package/integrations/haystack.py` (added in the
 * follow-on PR — see `docs/integrations/haystack.md`).
 *
 * Every `writeDocuments` call funnels through the MCOP triad
 * (encode → resonate → record → etch), so a Haystack pipeline gets
 * Merkle-rooted provenance for free. `filterDocuments` returns the
 * subset matching the supplied filter; `recallByResonance` exposes the
 * Stigmergy retrieval surface for callers who want resonance-ranked
 * search instead of metadata filtering.
 */

import { randomUUID } from 'node:crypto';
import {
  ensureTriad,
  recallFromTriad,
  recordIntoTriad,
  type MCOPProvenance,
  type MCOPTriad,
  type MCOPTriadOptions,
} from './triadHarness';

/** Haystack 2.x `Document` shape (framework-agnostic subset). */
export interface MCOPHaystackDocument {
  readonly id: string;
  readonly content: string;
  readonly meta?: Record<string, unknown>;
  readonly score?: number;
  readonly embedding?: ReadonlyArray<number>;
  readonly provenance?: MCOPProvenance;
}

/** Haystack 2.x `DuplicatePolicy` (subset). */
export type MCOPHaystackDuplicatePolicy = 'overwrite' | 'skip' | 'fail';

/** Haystack 2.x filter spec (subset — equality only, mirrors LangChain pattern). */
export type MCOPHaystackFilters = Readonly<Record<string, unknown>>;

export interface MCOPHaystackDocumentStoreOptions extends MCOPTriadOptions {
  readonly defaultPolicy?: MCOPHaystackDuplicatePolicy;
}

/**
 * MCOP-backed Haystack 2.x `DocumentStore` implementation.
 *
 * Public surface mirrors `haystack.document_stores.in_memory.InMemoryDocumentStore`
 * so a pipeline can swap a stock Haystack store for this shim without
 * other code changes.
 */
export class MCOPHaystackDocumentStore {
  private readonly triad: MCOPTriad;
  private readonly documents = new Map<string, MCOPHaystackDocument>();
  private readonly documentsByTraceId = new Map<string, string>();
  private readonly defaultPolicy: MCOPHaystackDuplicatePolicy;

  constructor(options: MCOPHaystackDocumentStoreOptions = {}) {
    this.triad = ensureTriad(options);
    this.defaultPolicy = options.defaultPolicy ?? 'overwrite';
  }

  /** Haystack `DocumentStore.count_documents`. */
  async countDocuments(): Promise<number> {
    return this.documents.size;
  }

  /**
   * Haystack `DocumentStore.write_documents`. Returns the number of
   * documents written (after applying the duplicate policy).
   */
  async writeDocuments(
    documents: ReadonlyArray<MCOPHaystackDocument>,
    policy?: MCOPHaystackDuplicatePolicy,
  ): Promise<number> {
    const effectivePolicy = policy ?? this.defaultPolicy;
    let written = 0;
    for (const doc of documents) {
      const existing = this.documents.get(doc.id);
      if (existing) {
        if (effectivePolicy === 'fail') {
          throw new Error(`Duplicate document id: ${doc.id}`);
        }
        if (effectivePolicy === 'skip') continue;
      }
      const recorded = recordIntoTriad(
        this.triad,
        doc.content,
        {
          ...(doc.meta ?? {}),
          mcop_haystack_document_id: doc.id,
        },
        `mcop-haystack:${doc.id}`,
      );
      const stored: MCOPHaystackDocument = {
        ...doc,
        embedding: doc.embedding ?? recorded.trace.context,
        provenance: recorded.provenance,
        meta: {
          ...(doc.meta ?? {}),
          mcop_stigmergy_trace_id: recorded.trace.id,
          mcop_etch_hash: recorded.etch.hash,
        },
      };
      this.documents.set(stored.id, stored);
      this.documentsByTraceId.set(recorded.trace.id, stored.id);
      written += 1;
    }
    return written;
  }

  /** Haystack `DocumentStore.filter_documents` — equality filter only. */
  async filterDocuments(
    filters?: MCOPHaystackFilters,
  ): Promise<ReadonlyArray<MCOPHaystackDocument>> {
    const all = Array.from(this.documents.values());
    if (!filters || Object.keys(filters).length === 0) return all;
    return all.filter((doc) => {
      const meta = doc.meta ?? {};
      return Object.entries(filters).every(([key, value]) => meta[key] === value);
    });
  }

  /** Haystack `DocumentStore.delete_documents`. */
  async deleteDocuments(documentIds: ReadonlyArray<string>): Promise<void> {
    for (const id of documentIds) {
      this.documents.delete(id);
      for (const [traceId, docId] of this.documentsByTraceId.entries()) {
        if (docId === id) {
          this.documentsByTraceId.delete(traceId);
        }
      }
    }
  }

  /**
   * Resonance retrieval — returns the best-matching document by Stigmergy
   * resonance against the query string, or `null` if nothing resonates.
   */
  async recallByResonance(
    query: string,
  ): Promise<{ score: number; document: MCOPHaystackDocument | null }> {
    const { resonance } = recallFromTriad(this.triad, query);
    if (!resonance.trace) return { score: resonance.score, document: null };
    const docId = this.documentsByTraceId.get(resonance.trace.id);
    const document = docId ? this.documents.get(docId) ?? null : null;
    return { score: resonance.score, document };
  }

  /** Expose the underlying triad for advanced callers. */
  get triadHandle(): MCOPTriad {
    return this.triad;
  }
}

/** Factory matching the rest of `src/integrations/`. */
export function createMCOPHaystackDocumentStore(
  options: MCOPHaystackDocumentStoreOptions = {},
): MCOPHaystackDocumentStore {
  return new MCOPHaystackDocumentStore(options);
}

/** Convenience: build a Haystack-shape document from raw content + metadata. */
export function mcopHaystackDocumentFromContent(
  content: string,
  meta?: Record<string, unknown>,
): MCOPHaystackDocument {
  return {
    id: randomUUID(),
    content,
    meta,
  };
}
